### ISSUE TITLE
Bug: jsxFactory causes no-unsafe-return and no-unused-vars

### ISSUE THREAD

**Issue by vbezhenar** [typescript-eslint/typescript-eslint#7151]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I'm trying to use jsxFactory feature to use jsx syntax for my own needs. TypeScript compiler seems to compile my code just fine, however eslint emits strange warnings and errors.

Here's relevant code and warnings:

```
/** @jsx myJsxFactory */

export function MyReport(): MyElement {
  return <MyPage />; // Unsafe return of an `any` typed value     @typescript-eslint/no-unsafe-return
}

function MyPage(): MyElement {
  return new MyElement();
}

function myJsxFactory(): MyElement { // 'myJsxFactory' is defined but never used  @typescript-eslint/no-unused-vars
  return new MyElement();
}

class MyElement {}

```

I tried to use `jsxFactory` in `tsconfig.json`, I also tried to set `parserOptions.ecmaFeatures: { jsx: true }` in `.eslintrc.json`, but it didn't change anything.

### Reproduction Repository Link

https://github.com/vbezhenar/tsx-eslint-issue

### Repro Steps

1) clone the repo
2) `npm install`
3) `npx eslint .`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.60.1` |
| `@typescript-eslint/parser`            | `5.60.1` |
| `@typescript-eslint/scope-manager`     | `5.60.1` |
| `@typescript-eslint/typescript-estree` | `5.60.1` |
| `@typescript-eslint/type-utils`        | `5.60.1` |
| `@typescript-eslint/utils`             | `5.60.1` |
| `TypeScript`                           | `5.0.4` |
| `ESLint`                               | `8.43.0` |
| `node`                                 | `v18.16.0` |

**Comment by Josh-Cena** [comment]
Unfortunately, TS-ESLint is not designed to handle JSDoc. You can send an issue like https://github.com/gajus/eslint-plugin-jsdoc/issues/858 asking eslint-plugin-jsdoc to add support for `@jsxfactory`.

**Comment by vbezhenar** [comment]
This issue has nothing to do with jsdoc. It's about typescript compiler jsxFactory option. Comment was there just do switch typescript option for the file. I updated the repo to reflect it. With jsxFactory option in tsconfig it doesn't work either.

**Comment by bradzacher** [comment]
So we definitely don't support the `@jsx` comment style - and this is entirely intentional because the usecases for it in modern code are ultimately few and far between. It's rare that someone wants a per-file factory and instead almost all usecases are covered via the top-level tsconfig option. Thus we only support it passed via the eslint config [`parserOptions.jsxPragma`](https://typescript-eslint.io/packages/parser#jsxpragma).

If you're doing a type-aware lint run we'll automatically fill that option from your tsconfig. You can double check this is happening by doing a lint run with the environment variable `DEBUG=typescript-eslint:*` and you'll see a log line that looks like this:
```
typescript-eslint:parser:parser Resolved jsxPragma from program: myJsxFactory +0ms
```

----

Now onto `no-unsafe-return`.
In your example code you think this is an error - however it's actually correct!
If you remove the explicit return type from `MyReport` you'll see that the inferred return type is `any`.
![image](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/db654b1d-28e6-4356-adb9-a634774a317f)

So TS is telling us the return type of the JSX expression is `any` and that's what we report on.
This is working as expected - if you fix your setup so that the JSX expression is not `any`, then the rule will not report on it.

----

Finally `no-unused-vars`.
This is a funny one and really is a big edge-case in our tooling. In a nutshell when our scope analyser encounters a JSX expression it will attempt to add a reference to the JSX factory.

However this is kind-of hackily implemented because the scope analysis tooling wasn't designed to work this way - so what it does is searches for variable in the upper scope with the name and adds a reference. However scope analysis is done via depth-first AST traversal.
Which means that in your example we see `<MyPage />` before we see `function myJsxFactory` - i.e. when our logic attempts to add a reference to `myJsxFactory`, the variable doesn't exist yet, and thus it is unable to add a reference.

This isn't a bug we've seen before because again - the usecases for custom JSX factories are one in a million and the usecases for custom JSX factories defined in the current file are one in a billion. Most usecases for JSX are plain react, some usecases are preact which require the custom config - but something else is super niche.

If you reorder your code so that `myJsxFactory` is defined *before* `MyReport` - you'll see the error is no longer reported.

This is definitely fixable - however it'll require a very large refactor of the scope analysis tooling as we'd need to essentially defer the JSX factory resolution until after a scope is "closed" and if the resolution fails, we'd need to register it against the parent scope and repeat.

Given that (a) the complexity this fix introduces is quite large and (b) the usecase of "a custom JSX factories with the JSX factory defined in the same file AND the JSX factory defined after the JSX expression" is infinitesimally rare (you're the first person in almost 4 years that's encountered this specific issue) - I'm okay with us not fixing this bug and leaving it as a known problem with the following workarounds:

1) reorder the function declarations so that the factory is defined before the first JSX expression
2) use a disable comment to disable the unused vars rule

**Comment by vbezhenar** [comment]
Thanks a lot for explanation and sorry for wasted time. I'll post fragments of my solution in case someone stumbles upon similar issue.

tsconfig.json:
```
  "compilerOptions": {
    "jsx": "react",
    "jsxFactory": "jsxFactory",
  }
```

```
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    type Element = MyElement;
  }
}

export function jsxFactory() {
  return new MyElement();
}

class MyElement {}

export function Text({ x, y }: { x: number; y: number }) {
  console.log(x, y);
  return new MyElement();
}

export function MyPage() {
  return (
    <Text x={100} y={100}>
      Hello
    </Text>
  );
}
```

This is just rough sketch which lints without warnings. Of course proper implementation of corresponding functions is required.

It compiles to

```js
export function jsxFactory() {
    return new MyElement();
}
class MyElement {
}
export function Text({ x, y }) {
    console.log(x, y);
    return new MyElement();
}
export function MyPage() {
    return (jsxFactory(Text, { x: 100, y: 100 }, "Hello"));
}
```
